torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -772,6 +810,57 @@ bool AliasDb::mayAlias(const Value* a, const Value* b) const {
   return memoryDAG_->mayAlias(elementMap_.at(a), elementMap_.at(b));
 }
 
+bool AliasDb::mayContainAlias(const Value* elem, const Value* container) const {
+  if (mayAlias(elem, container)) {
+    return true;
+  }
+
+  if (!shouldAnnotate(elem) || !shouldAnnotate(container)) {
+    return false;
+  }
+
+  if (!elementMap_.count(elem) || !elementMap_.count(container)) {
+    return false;
+  }
+
+  // don't know how to analyze containers other than tuple
+  if (isContainerType(elem->type())) {
+    if (elem->node()->kind() != prim::TupleConstruct) {
+      return true;
+    }
+    auto inp = elem->node()->inputs();
+    return std::any_of(inp.begin(), inp.end(), [&](const Value* v) {
```

Comment:
I don't understand why you're comparing aliasing on the contained types directly. Shouldn't `MemoryDAG` do that for you?

The role of AliasDb wrappers are to do wildcard/shouldAnnotate processing on the values before dispatching to MemoryDAG for the actual query. So in this case, I think you should recursively check for wildcards in the lhs/rhs containers, then dispatch directly to `MemoryDag::mayContainAlias` to do the rest.

Reply:
You made the comment that we could make the logic recursive, that is what this is doing. Memory Dag can't do it for me because there are certain types that we have to handle at the border, including Wildcards but also container types that we don't know how to analyze yet - everything but tupleconstruct.

Reply:
I meant that you should separate out the following two steps:
1. Special case handling: check if there are any wildcards or non-supported container types in the lhs and rhs container trees. Return true if the answer is yes.
2. If there are no special cases, dispatch the memoryDAG to answer the query normally.

Right now you're doing a variation of what you were doing before—recursive comparisons on both sides of contained value tree, with each comparison doing its own traversal of both subtrees (the memory DAG call).

Reply:
Do you know why the 
`
  if (!elementMap_.count(a) || !elementMap_.count(b)) {
    return false;
  }
` 
check exists in mayAlias?

Reply:
Not sure; if you can remove it would be good—the invariant should be that any value for which `shouldAnnotate` returns true should have an element.

Reply:
Yep that's what i thought. will try to remove.